File: mascii2/src/test/java/com/kastkode/mascii2/musicelements/KeysigTests.java
package com.kastkode.mascii2.musicelements;

import static org.junit.Assert.assertTrue;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;

import org.junit.jupiter.api.Test;

import com.kastkode.mascii2.Mascii2ApplicationTests;
import com.kastkode.mascii2.ParseResult;
import com.kastkode.mascii2.musicelements.KeySignature.AdHoc;
import com.kastkode.mascii2.util.MasciiUtil;

public class KeysigTests {

	@Test
	public void keysigencoding() {
		AdHoc key = KeySignature.B_FLAT.alterationsMap();
		int origBitMap = key.asBitmap();
		
		assertEquals(key.getAccidental('b').adjust, -1);
		assertEquals(key.getAccidental('e').adjust, -1);
		assertEquals(key.getAccidental('c').adjust, 0);
		
		assertTrue(KeySignature.E_FLAT.alterationsMap().asBitmap() != key.asBitmap());
		
		key.setAccidental('a', Accidental.FLAT);
		assertEquals(KeySignature.E_FLAT.alterationsMap().asBitmap(), key.asBitmap());
		
		//make sure original unaffected
		assertTrue(origBitMap == KeySignature.B_FLAT.alterationsMap().asBitmap());
	}

	@Test
	public void relativeMajor() {
		assertEquals('c', MasciiUtil.relativeMajorDegree('a'));
		assertEquals('d', MasciiUtil.relativeMajorDegree('b'));
		assertEquals('b', MasciiUtil.relativeMajorDegree('g'));
		assertEquals('g', MasciiUtil.relativeMajorDegree('e'));
		assertEquals('a', MasciiUtil.relativeMajorDegree('f'));
	}
	
	@Test
	public void keysigLookup() {
		assertEquals(KeySignature.C, KeySignature.getByDegree('c', Accidental.NATURAL, false));
		assertEquals(KeySignature.D_FLAT, KeySignature.getByDegree('d', Accidental.FLAT, false));
		assertEquals(KeySignature.F_SHARP, KeySignature.getByDegree('f', Accidental.SHARP, false));
		assertEquals(KeySignature.C_SHARP, KeySignature.getByDegree('a', Accidental.SHARP, true));
		assertEquals(KeySignature.B_FLAT, KeySignature.getByDegree('g', Accidental.NATURAL, true));
		assertEquals(KeySignature.E, KeySignature.getByDegree('c', Accidental.SHARP, true));
	}

	@Test
	public void alterationMemory() {
		String minuet = "{key: gm}\n\n"
				+ "g [b A] G b= A@ G a b | c B A G";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
		assertEquals(0, treb.get(3).getPitch() - treb.get(0).getPitch());
		assertEquals(1, treb.get(4).getPitch() - treb.get(1).getPitch());		
	}	

	@Test
	public void alterationMemory2() {
		String minuet = "{key: fm}\n"
				+ "a F | E= f E";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
		assertEquals(treb.get(2).getPitch(), treb.get(4).getPitch());
	}	
	
	@Test
	public void alterationMemory3() {
		String minuet = "{key: gm}\n"
				+ "g [b A] G b= A@ G a b | c B A G";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
		assertEquals(0, treb.get(3).getPitch() - treb.get(0).getPitch());
		assertEquals(1, treb.get(11).getPitch() - treb.get(7).getPitch());
		
	}	

	@Test
	public void accidentalKey() {
		String minuet = "{key: d@}\n"
				+ "f e d g";
		ParseResult result = Mascii2ApplicationTests.parse(minuet);
		assertEquals(
				KeySignature.D_FLAT.alterationsMap().asBitmap(), 
				result.getParts().get(0).getCurrentKey().asBitmap());		
	}	
}

File: mascii2-typescript/test/musicelements/KeysigTests.ts
import { test, describe } from 'node:test';
import assert from 'node:assert/strict';
import { SourceParser } from '../../src/SourceParser';
import { ParseResult } from '../../src/ParseResult';
import { Accidental } from '../../src/musicelements/Accidental';
import { KeySignature } from '../../src/musicelements/KeySignature';
import { relativeMajorDegree } from '../../src/util/MasciiUtil';

function parse(source: string): ParseResult {
    const result = new SourceParser().generateFromString(source);
    const msgs = result.errors?.getMsgs() ?? [];
    for (const msg of msgs) console.error(msg);
    assert.equal(msgs.length, 0, `Expected no parse errors, got: ${msgs.join(', ')}`);
    return result;
}

describe('KeysigTests', () => {

    test('keysigencoding', () => {
        const key = KeySignature.B_FLAT.alterationsMap();
        const origBitMap = key.asBitmap();

        assert.equal(key.getAccidental('b'), Accidental.FLAT);   // -1
        assert.equal(key.getAccidental('e'), Accidental.FLAT);   // -1
        assert.equal(key.getAccidental('c'), Accidental.NATURAL); // 0

        assert.notEqual(KeySignature.E_FLAT.alterationsMap().asBitmap(), key.asBitmap());

        key.setAccidental('a', Accidental.FLAT);
        assert.equal(KeySignature.E_FLAT.alterationsMap().asBitmap(), key.asBitmap());

        // make sure original unaffected
        assert.equal(origBitMap, KeySignature.B_FLAT.alterationsMap().asBitmap());
    });

    test('relativeMajor', () => {
        assert.equal('c', relativeMajorDegree('a'));
        assert.equal('d', relativeMajorDegree('b'));
        assert.equal('b', relativeMajorDegree('g'));
        assert.equal('g', relativeMajorDegree('e'));
        assert.equal('a', relativeMajorDegree('f'));
    });

    test('keysigLookup', () => {
        assert.equal(KeySignature.C,       KeySignature.getByDegree('c', Accidental.NATURAL, false));
        assert.equal(KeySignature.D_FLAT,  KeySignature.getByDegree('d', Accidental.FLAT,    false));
        assert.equal(KeySignature.F_SHARP, KeySignature.getByDegree('f', Accidental.SHARP,   false));
        assert.equal(KeySignature.C_SHARP, KeySignature.getByDegree('a', Accidental.SHARP,   true));
        assert.equal(KeySignature.B_FLAT,  KeySignature.getByDegree('g', Accidental.NATURAL, true));
        assert.equal(KeySignature.E,       KeySignature.getByDegree('c', Accidental.SHARP,   true));
    });

    test('alterationMemory', () => {
        const minuet = '{key: gm}\n\ng [b A] G b= A@ G a b | c B A G';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(0, treb[3]!.getPitch() - treb[0]!.getPitch());
        assert.equal(1, treb[4]!.getPitch() - treb[1]!.getPitch());
    });

    test('alterationMemory2', () => {
        const minuet = '{key: fm}\na F | E= f E';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(treb[2]!.getPitch(), treb[4]!.getPitch());
    });

    test('alterationMemory3', () => {
        const minuet = '{key: gm}\ng [b A] G b= A@ G a b | c B A G';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(0, treb[3]!.getPitch() - treb[0]!.getPitch());
        assert.equal(1, treb[11]!.getPitch() - treb[7]!.getPitch());
    });

    test('accidentalKey', () => {
        const minuet = '{key: d@}\nf e d g';
        const result = parse(minuet);
        assert.equal(
            KeySignature.D_FLAT.alterationsMap().asBitmap(),
            result.getParts()![0]!.currentKey.asBitmap(),
        );
    });

});
